### ISSUE TITLE
Mark alpha/beta versions and show stable alternatives

### ISSUE THREAD

**Issue by Xerus2000** [ben-manes/gradle-versions-plugin#271]
Right now, it is sometimes not obvious that the latest version offered by the plugin is an alpha or beta, so I have to browse the maven repository to check that and find out the latest stable version. This relates to #69, but it'd be nice if you could also use color to make alphas stand out and additionally show the latest stable version.

**Comment by ben-manes** [comment]
The version number is all that we know. Doesn’t that indicate it’s release status? Can you provide an example?

**Comment by Xerus2000** [comment]
Yeah that's true, it is usually obvious. But the main issue is that I don't get the latest stable version. Does maven have an alpha/beta status or does it just infer it from keywords (like alpha/beta/rc)?

**Comment by ben-manes** [comment]
It has [conventions](
https://maven.apache.org/pom.html#Version_Order_Specification), but few follow them. We can only allow you to filter using a resolution strategy. Even then it’s not perfect, e.g. some projects use even/odd to denote stability.

**Comment by marcardar** [comment]
Out of interest, when the output says `com.android.tools.build:gradle [3.3.0-rc03 -> 3.4.0-alpha10]` does this mean I can be sure that a later `3.3.0` release candidate has not yet been released? If it does not imply that, then is it possible to configure this plugin so that it does give something like `[3.3.0-rc03 -> 3.3.0-rc04]` even when a later alpha is available?

**Comment by ben-manes** [comment]
We let Gradle perform the version comparison order, via its [StaticVersionComparator](https://github.com/gradle/gradle/blob/52fe778f987608aba01bb4ecb7dde27ad76f3c0a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/strategy/StaticVersionComparator.java). That should conform to Maven's [Version Order Specification](https://maven.apache.org/pom.html#Version_Order_Specification), but I don't see special handling for `alpha` so it may not. Any improvements there should to requested on Gradle's issue tracker.

You can use a resolution strategy to reject candidate versions, like those containing `alpha`. See the README for examples.

**Comment by marcardar** [comment]
Thanks, yes I just saw the README, and that's very useful. My thinking here, is that, from alpha/beta/rc I want to know about updates towards stability. From stable releases, I want to know about updates to other stable releases. Anything moving from something to a less stable release I would only do because of some new feature and that new feature would've already made me aware of that new release, so, for me, this tool would not be needed for that.

**Comment by ben-manes** [comment]
Maven's versioning strategy, and resulting metadata, never accounted for that detail. It is merely a list of versions, with the most recently created as the latest. They expected `SNAPSHOT` versions to be used for development versions, but that serves a different purpose. Since many only casually follow Maven's rules and have their own conventions, it gets very murky. Instead of rewriting all of Gradle's dependency management logic, we script it so any limitations or improvements are inherited.

The best we could offer might be #69, but that doesn't solve your case either.

If you maintain your dependencies in a well known format, e.g. like a [config file](https://github.com/ben-manes/caffeine/blob/master/gradle/dependencies.gradle), you could query that it compare against your latest. Then decide in your resolution strategy if the candidate is acceptable.

**Comment by guenhter** [comment]
@Xerus2000 Doesn't the resolution strategy provided in the README solves this?

```
tasks.named<DependencyUpdatesTask>("dependencyUpdates") {
  resolutionStrategy {
    componentSelection {
      all {
        val rejected = listOf("alpha", "beta", "rc", "cr", "m", "preview")
          .map { qualifier -> Regex("(?i).*[.-]$qualifier[.\\d-]*") }
          .any { it.matches(candidate.version) }
        if (rejected) {
          reject("Release candidate")
        }
      }
    }
  }
  // optional parameters
  checkForGradleUpdate = true
  outputFormatter = "json"
  outputDir = "build/dependencyUpdates"
  reportfileName = "report"
}
```

Without this strategy, I get e.g.

```
- ch.qos.logback:logback-classic [1.2.2 -> 1.3.0-alpha4]
```


but with the strategy enabled, it is:

```
- ch.qos.logback:logback-classic [1.2.2 -> 1.2.3]
```

**Comment by Xerus2000** [comment]
But I don't want to hide them, I just want them to be immediately apparent as being unstable versions.

**Comment by ben-manes** [comment]
If we had #69, do you think the rest is just making the report pretty? As in, we'd have all of the limitations of Gradle but for those following semi-standard patterns we could resolve an upgrade chain. At which point a custom report could render it however it sees fit, e.g. using colors based on a keyword match as you suggested.

**Comment by Xerus2000** [comment]
Yeah, I guess that would suffice. Maybe you could do some pretty rendering by default and have a config option to turn it off ^^
